### PR TITLE
Cache the result of type-related queries (specifically for PSQL)

### PR DESCRIPTION
**Pull request by ted-hanson**
### Summary

SchemaCache interface heavily based off PR #39077 but with a generic cache interface + a specific one for mysql, psql, and sqlite3

The major purpose of this PR is that using the schema cache file (generated by rake db:schema:cache:dump) would also cache the result of type-related queries for PostgreSQL, so it will reduce 2 queries for each new database connection. As a side effect, it will also reduce 2 queries for checking-out new PostgreSQL connection within a connection pool.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb [hunk 1]
```diff
@@ -49,21 +49,21 @@ def initialize_dup(other)
       def encode_with(coder)
         reset_version!
 
-        coder["columns"]          = @columns
-        coder["primary_keys"]     = @primary_keys
-        coder["data_sources"]     = @data_sources
-        coder["indexes"]          = @indexes
-        coder["version"]          = @version
-        coder["database_version"] = database_version
+        coder["columns"]                  = @columns
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb:52]: What is the reason for this change?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb [hunk 2]
```diff
@@ -166,7 +166,6 @@ def dump_to(filename)
 
       def marshal_dump
         reset_version!
-
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb:169]: Please put the line back.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 3]
```diff
@@ -41,6 +41,10 @@ class AbstractAdapter
 
       attr_accessor :pool
       attr_reader :visitor, :owner, :logger, :lock
+
+      cattr_accessor :additional_type_records_cache, default: {}
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:45]: This is only used in the postgresql adapter, why it is being defined in the abstract adapter?

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 4]
```diff
@@ -54,6 +55,10 @@ def initialize(connection, logger, connection_options, config)
         super(connection, logger, config)
       end
 
+      def init_schema_cache
+        MySQL::SchemaCache.new(self)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:59]: Why are we defining a special schema cache for MySQL? I think we don't need that.

**Comment by ted-hanson** [comment]
@rafaelfranca thanks for the comments + review! I'll definitely look into them as I dig through it after hours.  It's been awhile so I have to get in the PR mindset again.

**Comment by beauraF** [comment]
@ted-hanson I'm very interested by this PR. How can I help you on this?

**Comment by ted-hanson** [comment]
@beauraF honestly got caught up in work and forgot to make changes, I can try to bang this out this weekend

**Comment by beauraF** [comment]
That would be amazing. It's a big pain point for my company, and the cost of maintaining our monkey patch is high. I can try to help.

**Comment by piecehealth** [comment]
@ted-hanson I suggest that you can only create a PostgreSQL `SchemaCache` rather than touch other adapters since the type related things are very PostgreSQL specific. You can refer to the original PR https://github.com/rails/rails/pull/39821, I have already done the changes.

@beauraF if you need this feature, you can refer to https://github.com/opf/openproject/pull/8563 for the monkey patch approach. Our application also has a monkey patch for this feature, please let me know if there is anything I can help.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb (hunk 1 above)
Comment (ted-hanson) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_cache.rb:52]: linter just did this automatically
